dynamodb/schema.go

```diff
@@ -0,0 +1,339 @@
+	// A map from column name to a count map of possible data types.
+	stats := make(map[string]map[string]int64)
+	var lastEvaluatedKey map[string]*dynamodb.AttributeValue
+	var count int64
+
+	for {
+		// Build the query input parameters
+		params := &dynamodb.ScanInput{
+			TableName: aws.String(s.TableName),
+		}
+		if lastEvaluatedKey != nil {
+			params.ExclusiveStartKey = lastEvaluatedKey
+		}
+
+		// Make the DynamoDB Query API call
+		result, err := client.Scan(params)
+		if err != nil {
+			return nil, fmt.Errorf("failed to make Query API call for table %v: %v", s.TableName, err)
+		}
+
+		// Iterate the items returned
+		for _, attrsMap := range result.Items {
+			for attrName, attr := range attrsMap {
+				if _, ok := stats[attrName]; !ok {
+					stats[attrName] = make(map[string]int64)
+				}
+				incDyDataTypeCount(attrName, attr, stats[attrName])
+			}
+
+			count++
+			if count >= sampleSize {
+				return stats, nil
+			}
+		}
+		if result.LastEvaluatedKey == nil {
+			break
+		}
+		// If there are more rows, then continue.
+		lastEvaluatedKey = result.LastEvaluatedKey
+	}
+	return stats, nil
+}
+
+func incDyDataTypeCount(attrName string, attr *dynamodb.AttributeValue, s map[string]int64) {
+	if attr.S != nil {
+		incCount(s, typeString)
+	} else if attr.BOOL != nil {
+		incCount(s, typeBool)
+	} else if attr.N != nil {
+		if int64Parsable(*attr.N) {
+			incCount(s, typeNumberInt)
+		} else {
+			incCount(s, typeNumberFloat)
+		}
+	} else if len(attr.B) != 0 {
+		incCount(s, typeBinary)
+	} else if attr.NULL != nil {
+		// Skip because all optional attributes are nullable.
+	} else if len(attr.L) != 0 {
+		incCount(s, typeList)
+	} else if len(attr.M) != 0 {
+		incCount(s, typeMap)
+	} else if len(attr.SS) != 0 {
+		incCount(s, typeStringSet)
+	} else if len(attr.NS) != 0 {
+		if int64Parsable(*attr.NS[0]) {
+			incCount(s, typeNumberIntSet)
+		} else {
+			incCount(s, typeNumberFloatSet)
+		}
+	} else if len(attr.BS) != 0 {
+		incCount(s, typeBinarySet)
+	} else {
+		log.Printf("Invalid DynamoDB data type: %v - %v", attrName, attr)
+	}
+}
+
+func (s *dynamoDBSchema) inferDataTypes(stats map[string]map[string]int64) {
+	type statItem struct {
+		Type  string
+		Count int64
+	}
+
+	if s.ColumnTypes == nil {
+		s.ColumnTypes = make(map[string]string)
+	}
+
+	for col, countMap := range stats {
+		var statItems []statItem
+		for k, v := range countMap {
+			statItems = append(statItems, statItem{Type: k, Count: v})
+		}
+
+		if len(statItems) == 0 {
+			log.Printf("Skip empty column %v", col)
+			continue
+		}
+
+		// Sort the slice reversely so the most frequent data type will be
+		// placed first.
+		sort.Slice(statItems, func(i, j int) bool {
+			// If counts are equal, then sort by names in alphabetical order.
+			if statItems[i].Count == statItems[j].Count {
+				return statItems[i].Type < statItems[j].Type
+			}
+			return statItems[i].Count > statItems[j].Count
+		})
```

Comment:
I'm wondering if this heuristic is the right one.
For example, if we have 60% Float and 40% String would we want to use Float?
For some low-levels of conflicting types, we might want to consider them errors or random spurious events.
But if there is a significant level of conflicting types, we might want to consider using String.